ls1intum/Artemis
# Refactor submissions counting for tutor/instructor dashboards

### Checklist

### Motivation and Context


We want to reduce code duplication and requests to the database when getting data for tutor and instructor dashboards.

### Description

The query methods to count number of submissions were moved from it's respective repositories and services of concrete exercise types (modeling, text, file-upload) to `SubmissionRepository`. This allowed to avoid checking for concrete exercise type and doing multiple calls for each exercise type. Also use more primitive types to avoid casting in Course and Exercise resources.

**Test Coverage**
Both `CourseResource` and `ExerciseResource` that were changed by this PR are covered by existing tests. I made small adjustments to them. Missed branches were not touched by this PR and reflect mostly exceptional cases.
![Screenshot_2020-03-05 de tum in www1 artemis web rest(5)](https://user-images.githubusercontent.com/17262376/76029014-fe248b00-5f33-11ea-8a3f-f352f73dc853.png)
![Screenshot_2020-03-05 de tum in www1 artemis web rest(4)](https://user-images.githubusercontent.com/17262376/76028771-8b1b1480-5f33-11ea-9d7c-a38a47e6ebc2.png)

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Select Tutor Dashboard 
4. Check that all numbers are correct on the dashboard
5. Navigate to exercise and also check the numbers are correct
6. Do the same for Instructor Course and Exercise Dashboards
### Screenshots


![Screenshot_2019-12-19 Tutor Exercise Dashboard](https://user-images.githubusercontent.com/17262376/71216376-a5d9d780-22ba-11ea-9c93-4dc92122f1c2.png)
---
![Screenshot_2019-12-19 Instructor Course Dashboard](https://user-images.githubusercontent.com/17262376/71216389-a96d5e80-22ba-11ea-9caa-11eeba3a4a9b.png)
---
![Screenshot_2019-12-19 Instructor Exercise Dashboard](https://user-images.githubusercontent.com/17262376/71216396-abcfb880-22ba-11ea-8240-757bff3d6911.png)
